Docs/adr fungibility (#319)
allinbits/cosmos-cash: docs/Explanation/ADR/adr-009-fungibility.md

# ADR 009: Fungibility

## Changelog

- 2021-12-03: Initial draft.
- 2021-12-05: Updated draft.

## Status

PROPOSED

## Abstract

Cosmos Cash is a platform enabling the issuance of MICA-compliant e-money tokens.

Two approaches to e-money token Fungibility are outlined.

> Cosmos Cash supports the second approach.

Reviewer: This abstract should outline what the two options are.

## Context

###EMT:

EMT = e-money token.

###Fungibility:

Fungibility is defined as an absolute, 2 EMTs of the same numeraire are Fungible if they are absolutely interchangeable; as a result, they have the same value.

> ###Fungibilty Approach 1:

Reviewer: This would benefit from an example - so USDC USDT TrueUSD etc.

Under this approach, the chain will have multiple EMTs - per numeraire.

Suppose n issuers issue EMTs.

The issuers do not support Fungibility, for example by agreeing to exchange and redeem the EMTs of other issuers.

EMTs are not Fungible under Approach 1.

It is believed that Fungibility Approach 1 gives a poor user experience, akin to a high street where each shop accepts different EUR notes and where
shoppers' wallets each contain a subset of the different available EUR notes.

> ###Fungibilty Approach 2:

Reviewer: Would this benefit from an example? e.g. USDC

Under this approach, the chain will support one EMT - per numeraire.

> Suppose n issuers issue EMTs.

Reviewer: What does this mean in terms of the discussion?

The issuers agree to support Fungible (at par) exchange and redemption of EMTs from other issuers.

Because they can always readily exchange EMTs from different issuers at par, users will treat EMTs from different Issuers as homogeneous.

EMTs are Fungible under Approach 2.

It is believed that Fungibility Approach 2 provides 2 benefits:
* An improved user experience.
* A network effect, increasing EMT utility (i.e. if more users accept/use a given EMT, the value from its use and, also its liquidity are increased).

> Both benefits ultimately lead to greater EMT use, EMT issuance and, chain economic value.

Reviewer: How? Why?

> Under Fungibility Approach 2, if and when an issuer holds more of another issuer's EMTs than it holds of theirs, it does face Counterparty Risk. It can mitigate this by redeeming competitor EMTs ASAP.

Reviewer: I think might need explaining or exploring more. For example, EMT legislation means tokens are backed one-2-one by underlying that can be redeemed almost immediately.
An example is how USDC is a consortium so that means some governance around issuers - presumably to mitigate counterparty risk and set standards on issuance. What does this mean?

## Decision

Cosmos Cash will support the second approach.

* The chain will support one EMT - per numeraire.

* The chain's EMT will have a chosen name, such as EUR-E.

* Issuers support Fungibility on chain:
   - A transaction type is available as follows: for User U, licensed by Issuer I1, U can 'send' to I1 the EMT an Issuer I2, 'receiving' in returning the EMT of I1.
   - The transaction will fail if: I1 is up to its issuance limit or the Issuer license of I2 has been revoked.

## Consequences

See Further Discussions.

### Backwards Compatibility

This ADR is expected to limit Backward Compatibility, but this compatibility remains to be confirmed.

### Positive

> N/A.

Reviewer 2: Could we add here:

- issuers can easily have interoperable stable tokens without the need for complex swapping mechanisms
- cosmos cash protocol will be easier to build and reason about as there will be a less total number of tokens

wdyt?

### Negative

> N/A.

Reviewer 2: Could we add here:

- issuers will find it more difficult to white label their tokens

wdyt?

# ADR 009: Fungibility

## Changelog

- 2021-12-03: Initial draft

## Status

PROPOSED

> For details on ADR workflow, see the [PROCESS](./PROCESS.md#adr-status) page.
> Use DRAFT if the ADR is in a draft stage (draft PR) or PROPOSED if it's in review.

## Abstract

It is anticipated that Elesto will have more than one Issuer issuing e-money tokens denominated in any given currency (numereraire).

Here we summarise the anticipated behaviour and driving incentives of Issuers and Users and, what this means for our choice of functionality.

## Context

For simplicity of description, assume throughout we are discussing USD-denominated tokens.

EMT = e-money token.

> Issuer behaviour:

Author: Uodated.

Issuers agree to support Fungible (at par) exchange and redemption of EMTs from other currently licensed issuers.

Issuer incentives for the given behaviour:

(a) Benefit: network effect leads to increased EMT use, with increased issuance and circulation.
    (This is driven by the resulting User behaviour and incentives - below. )
(b) Benefit: any given issuer would lose market share if they didn't agree to support fungility as stated.
    (Exchange and redemption take their competitors' EMTs out of circulation, put their EMT into circulation.)
(c) Cost: Issuers face Counterparty risk only while they hold competitors' EMTs - they can mitigate this by redeeming competitor tokens.

User behaviour:

(i) Users will treat EMTs from different Issuers are homogeneous.
(ii) Users will accept wallets and related use cases that equally 'handle' EMTs of different Issuers as homogenous.

>      This will lead to enhanced 'ease of use' and, further adoption (see: Issuer incentives).

Author: Updated.

## Decision

> We will update our functionality as follows:
> 
> (1) Our chain will support only one EMT (per currency), issued by 1 or more Issuers.
> 
> (2) Our chain's EMTs will all have the same given name? (Such as USD-E for USD.)
> 
> (3) Added Issuer functionality.
>     We add a transaction type as follows:
>         For User U, licensed by Issuer I1, U can 'send' to I1 the EMT an Issuer I2, 'receiving' in returning the EMT of I.
>     The transaction will fail if:
>         I1 is up to its issuance limit or
>         The Issuer license of I2 has been revoked.

Reviewer 3: ```suggestion
Functionality will be updated as follows:

(1) Our chain will support only one EMT (per currency) that is issued by 1 or more Issuers.

(2) Our chain's EMTs will all have the same given name? For example, USD-E for USD.

(3) Added Issuer functionality to add a transaction type as follows:

    For User U, licensed by Issuer I1: 
    
    - U can 'send' to I1 the EMT an Issuer I2
    - U 'receiving' in returning the EMT of I

    The transaction will fail if:
    
    - I1 is up to its issuance limit or
    - The Issuer license of I2 has been revoked
```

Reviewer 3: check line 57... is it U that receives? 

I suggest using the clear list item formatting as suggested here so that the functionality is more easily scanned and understood

Author: Adopted.

## Consequences

See Context.

> ### Backwards Compatibility
> 
> Expected to limit Backwards Compatibility but, to be confirmed.

Reviewer 3: ```suggestion
### Backward Compatibility

Expected to limit Backward Compatibility, but this compatibility remains to be confirmed.
```

Author: Adopted.

### Positive

See Context.

### Negative

See Context.

### Neutral

See Context.

## Further Discussions

It is expected users will more readily adopt Fungible EMTs.

> The decision given is to only support Fungible EMTs.
> 
> Alternative Decision options are:
> (a) Only support non-Fungibile EMTs,
> (b) Support Fungible and non-Fungible EMTs
>     (For example Issuers I1 and I2 issue Fungible EMT T1 and, Issuer I3 issues non-Fungible EMT T2.)

Reviewer 3: ```suggestion
The decision given is to only support Fungible EMTs.

Alternative Decision options are:

(a) Support only non-Fungibile EMTs
(b) Support Fungible and non-Fungible EMTs. 

    For example, Issuers I1 and I2 issue Fungible EMT T1 and, Issuer I3 issues non-Fungible EMT T2.
```

Reviewer 3: I'm not clear on the preference for (a) vs a bullet

Author: Have moved to bullet.

## Test Cases [optional]

N/A

## References

> https://www.notion.so/allinbits/FungibilityV2-1-e6b1deed13cc46649e9f915d4a23944b

Reviewer 4: remove the link since it is private and not reachable by a broader audience. Perhaps link to other research like the USDC one?
```suggestion
```

Author: agreed
